draft-ietf-quic-tls.md

```diff
@@ -580,6 +661,185 @@ Prior to TLS providing keys, no record protection is performed and the
 plaintext, P, is transmitted unmodified.
 
 
+## Packet Numbers {#packet-number}
+
+QUIC has a single, contiguous packet number space.  In comparison, TLS
+restarts its sequence number each time that record protection keys are
+changed.  The sequence number restart in TLS ensures that a compromise of the
+current traffic keys does not allow an attacker to truncate the data that is
+sent after a key update by sending additional packets under the old key
+(causing new packets to be discarded).
+
+QUIC does not assume a reliable transport and is required to handle attacks
+where packets are dropped in other ways.  QUIC is therefore not affected by this
+form of truncation.
+
+The packet number is not reset and it is not permitted to go higher than its
+maximum value of 2^64-1.  This establishes a hard limit on the number of packets
+that can be sent.
+
+Some AEAD functions have limits for how many packets can be encrypted under the
+same key and IV (see for example {{AEBounds}}).  This might be lower than the
+packet number limit.  An endpoint MUST initiate a key update ({{key-update}})
+prior to exceeding any limit set for the AEAD that is in use.
+
+TLS maintains a separate sequence number that is used for record protection on
+the connection that is hosted on stream 1.  This sequence number is not visible
+to QUIC.
+
+
+# Key Phases
+
+As TLS reports the availability of 0-RTT and 1-RTT keys, new keying material can
+be exported from TLS and used for QUIC packet protection.  At each transition
+during the handshake a new secret is exported from TLS and packet protection
+keys are derived from that secret.
+
+Every time that a new set of keys is used for protecting outbound packets, the
+KEY_PHASE bit in the public flags is toggled.
+
+Once the connection is fully enabled, the KEY_PHASE bit allows a recipient to
+detect a change in keying material without necessarily needing to receive the
+first packet that triggered the change.  An endpoint that notices a changed
+KEY_PHASE bit can update keys and decrypt the packet that contains the changed
+bit, see {{key-update}}.
+
+The KEY_PHASE bit on the public flags is the most significant bit (0x80).
+{{key-phase-table}} summarizes the different values for the KEY_PHASE bit.
+
+| Scenario | Client (No 0-RTT) | Client (0-RTT) | Server |
+|:-|-:|-:|-:|
+| Handshake | 0 | 0 |  0 |
+| 0-RTT Data | - | 1 | - |
+| 1-RTT Data | 1 | 0 | 1 |
+| 1st Key Update | 0 | 1 | 0 |
+| 2nd Key Update | 1 | 0 | 1 |
+{: #key-phase-table title="Summary of KEY_PHASE Values"}
+
+{{key-phase-table}} shows that a client marks 1-RTT with a different KEY_PHASE
+bit depending on whether 0-RTT is attempted.  Attempting 0-RTT therefore results
+in fully protected data having different KEY_PHASE values.  This is true even if
+0-RTT data is rejected and ignored by the server.
```

Comment:
Does this work?  If the 0RTT packets are never received by the server, wouldn't 0RTT with packet loss and 1RTT be identical from the server's perspective, making it look like 0 for handshake and 0 for 1RTT data, with no 1 in between.

I think the approach of using the version bit, possibly requiring the version bit to be set for handshake packets, then requiring it not be set for 0RTT data, would be a reasonable option?  There are few ways to make it work, but the above doesn't seem like it works, and generally seems unnecessarily complex.